### DISCUSSION TITLE
Power Limiter not working?

### DISCUSSION THREAD

**Discussion by papageier** [hoylabs/OpenDTU-OnBattery#227] [Q&A]
Hey,

I managed to get the project to run on my ESP32, instead of the vanilla openDTU running there before. All seems well, communication is fine, I can control my Hoymiles (HM-300) as usual. Now, I wanted to use the Dynamic Power Limiter, in the most basic of all scenarios: just use the Start/Stop values vs voltage measured at the Hoymiles DC side, to cut-off AC production when the battery is low. My battery is a Pylontech (2x US2000C), so 50V vs 49V seems a good start. However, I can't get the Power Limiter to work at all. I played with all the options in the Dynamic Power Limiter settings panel. The inverter keeps producing, no matter the voltage on the Hoymiles. I activated debugging (uncommented the #ifdef in PowerLimiter.cpp) to see if the class would show up in the virtual console - but no luck, either. Looks like the PowerLimiter isn't working at all.

Is anyone using a similar setup and has experienced the same phenomenon? Or has an idea what I might be doing wrong?

My settings are:
- Power Limiter Enable: True
- Enable Pass Through: False
- Inverter ID: 0
- Inverter is behind power meter: True
- DC Voltage Start: 50
- DC Voltage Stop: 49
- Load Correction Factor: 0.001

Any hints (also in German) are very much appreciated! :-)

Regards,
Thomas

**Comment by MalteSchm** [comment]
Ist ein Power Meter eingestellt und welche Spannung hat die Batterie? (Gemessen und in dem Webinterface angezeigt)

**Reply by papageier** [reply]
Sorry für die späte Rückmeldung. Ich bin inzwischen ein Stück weiter (Code & Log lesen). Es sieht so aus, dass der PowerLimiter grundsätzlich ein PowerMeter erwartet, das regelmäßig Daten liefert. Ohne PowerMeter funktioniert er gar nicht, ohne aktuelle Daten schaltet er den Inverter ab, aus Sicherheitsgründen. Ich hatte zunächst kein PowerMeter konfiguriert (ich hab ja auch keins), damit kam der Code erst gar nicht in die Loop vom PowerLimiter. Jetzt habe ich mir einen (dummy) MQTT-PowerMeter konfiguriert, der aber natürlich keine Daten liefert --> PL schaltet den Inverter ab. Wenn ich das richtig verstehe, wäre das aber eigentlich - zumindest aus Sicherheitsgründen - gar nicht nötig, weil später in der Loop auf jeden Fall noch auf das Erreichen der unteren Ladungsgrenze der Batterie bzw. der unteren Spannungsgrenze am DC-Eingang geprüft wird. Spätestens dann würde der Inverter ja auch wieder abgeschaltet --> kann eigentlich nix passieren. Nur der Teil der Loop, die vom PowerMeter Daten erwartet, um dynamisch die Leistung zu steuern, die braucht natürlich aktuelle Daten vom PowerMeter.
Ich könnte mir vorstellen, einen neuen PowerMeter a la "Static: Battery State / DC Voltage" zu definieren, der das Ein- und Ausschalten des Inverters an den besagten Grenzwerten nach wie vor unterstützt, aber eben ohne externes PowerMeter auskommt. Das wäre dann eher statische Leistungsschaltung an den eingestellten Ober-/Untergrenzen. Genau das wäre aber mein Anwendungsfall: Ich will eigentlich nur, dass der Inverter sich abschaltet, sobald die Batterie einen unteren Ladezustand erreicht hat, und sich anschaltet, sobald sie wieder ausreichend geladen ist. Dann kann ich die Anlage nämlich laufen lassen und muss mich nicht um den Ladezustand kümmern.

Wenn jetzt meine C-Kenntnisse noch gut genug wären, gäb's an dieser Stelle auch den passenden Pull Request dazu. Aber dafür müsste ich zumindest ein bisserl Code reorganisieren, und bei Leistungssteuerungen macht das besser jemand, der weiß, was er da tut.

**Comment by helgeerbe** [comment]
OK, jetzt verstehe ich deinen Anwendungsfall. Und dafür ist der PowerLimiter eigentlich nicht gedacht. Über den Power Meter soll ja eine Nulleinspeisung realisiert werden.

Dein Anwendungsfall ist ja nur ein einfacher Batteriewächter. Wenn du kein BMS an deiner Batterie hast, würde ich mal überlegen, ob du dir nicht besser einen Victron Smart Battery Protect nimmst und ihn zwischen Batterie und Wechselrichter klemmst. Da ist zwar kein von Victron vorgesehener Einsatz (es wird im Handbuch auch explizit darauf hingewiesen), funktioniert aber zumindest am HM-300 trotzdem (auf eigenes Risiko). Ansonsten musst du immer davon ausgehen, dass die OpenDTU und der Wechselrichter immer richtig funktionieren. OpenDTU-onBattery ist ein Bastel- und Spaßprojek und kann weder für fehlerfreie Hardware und Software garantieren. 

Wenn ich du wäre, würde ich meine teure Batterie auf alle Fälle mit ein wenig mehr Aufwand schützen. Dann ist im Fehlerfall höchsten die Batterie leer. Aber nicht kaputt.

**Comment by MalteSchm** [comment]
Sehe das ähnlich. Ohne BMS sollte man so etwas nicht betreiben.

Ansonsten kannst du auch versuchen den "max_power" Wert deines Inverters als Power Meter Value zu nutzen. Der ist unter ..../api/limit/status abrufbar. Dazu konfigurierst du ein Power Meter der die API deines openDTU abfragt. Deine Leistung im Haus entspricht immer der Maximalleistung des Inverters. 

Alles was du dann noch tun musst ist den max Wert im Power Limiter auf deine gewünschte Leistung zu setzen.

Dann regelt der Power Limiter auf den max Wert solange der Batterieladezustand  das hergibt

**Reply by papageier** [reply]
Cool, klingt wie ungefähr das, was ich haben will. Probiere ich aus.
Meine Pylontechs haben schon ein BMS, sonst wäre mir das Konstrukt auch zu heikel. Sagt ja niemand, dass die OpenDTU nicht mal ausfällt oder das Netz oder sonstwas schief geht. Ich will die Batterie nur nicht regelmäßig bis in den Bereich herunter fahren, wo das BMS abriegelt - dann ist nämlich ggf. die ganze Anlage aus und meine Systemüberwachung schlägt Alarm. Ich würde einfach gern so bei 10-15% Batteriestatus den Inverter abschalten.

**Reply by madmartin** [reply]
Wenn Du Pylontech Batterien hast, warum schliesst Du nicht den CANbus an? Welches Modell hast Du?

**Reply by papageier** [reply]
Ich habe 2x US2000C, die per CAN-Bus mit einem Victron Cerbo GX reden, der wiederum per VE.direct mit dem MPP SmartSolar 100 redet. CAN-Bus-Port an der Batterie ist also belegt, ich könnte aber VE.direct vom Cerbo zur DTU verwenden. Nur - bringt mir das was? Dann liest die Batterie die Spannung aus, rechnet sie in einen %-Ladezustand um und schickt den an die OpenDTU. Die macht damit aber (wenn ich den Code richtig lese) nichts weiter, als den Inverter an den entsprechenden Ladezustands-Grenzen an und aus zu schalten. Das tut sie aber ohne die Verbindung zum BMS jetzt auch schon, direkt mit der DC-Eingangsspannung am Inverter, nur dass ich in diesem Fall die Spannung selber in den entsprechenden Ladezustand umrechnen muss. Oder entgeht mir da was?
